src/corner-kick/src/service/RosConnector.ts

```diff
@@ -0,0 +1,132 @@
+ * Singleton.
+ * 
+ * Used by TopicListener to handle incoming messages.
+ */
+export class ROSConnector {
+
+    private static instance: ROSConnector;
+
+    private ros: any; // TODO: create typing information for the ROS object
+
+    /**
+     * Stores the topics we are currently registered with. This helps us avoids subscribing to a topic we are already subscribed with.
+     */
+    private topics: string[] = [];
+
+    /**
+     * Stores the callback functions for each topic. When a new message arrives for a given topic, we call the functions stored in
+     * this dictionary.
+     */
+    private topicsSubscriptions: {[topic: string]: [(topic: any) => void]} = {};
+    
+
+    /**
+     * Returns an instance of ROSConnector
+     */
+    public static get Instance() {
+        return this.instance || (this.instance = new this());
+    }
+
+    private constructor() {
+        // Wait a little bit before connecting to ROS. It can sometimes take time
+        // to initialize.
+        setTimeout(() => {
+            this.ros = new ROSLib.Ros({
+                url: ROS_URL,
+            });
+
+            // TODO: handle connection errors
+
+            // We are connected, subscribe to ROS
+            this.subscribeToROS();
+        }, 1000);
+    }
+
+    /**
+     * Subscribe to a given topic. When new messages are received on the topic, the callback will be called.
+     * @param topic the topic to subscribe to
+     * @param callback the callback function to call when a new topic is received
+     */
+    public subscribeToTopic = (topic: string, callback: (topic: any) => void) => {
+        // We initialize a new array if this is the first callback for a given topic
+        if(this.topicsSubscriptions[topic] === undefined) {
+            this.topicsSubscriptions[topic] = [callback];
+        } else {
+            this.topicsSubscriptions[topic].push(callback);
+        }
+    }
+
+    /**
+     * Unsubscribe from a given topic.
+     * @param topic the topic to unsubscribe from
+     * @param callback the callback function to remove
+     */
+    public unsubscribeToTopic = (topic: string, callback: (topic: any) => void) => {
+        if(this.topicsSubscriptions[topic] !== undefined) {
+            const index = this.topicsSubscriptions[topic].indexOf(callback);
+            this.topicsSubscriptions[topic].splice(index, 1);
+        }
+    }
+
+    public publish = (name: string, messageType: string, message: any) => {
+        const topic = new ROSLib.Topic({
+            messageType,
+            name,
+            ros: this.ros,
+        });
+
+        topic.publish(message);
+    }
+
+    /**
+     * Internal method to subscribe to all ROS topics. This is the method that actually "subscribes" in a ROS sense. We do this to create an abstraction
+     * layer between ROS and the rest of our application
+     */
+    private subscribeToROS = () => {
```

Comment:
If this "subscribes to all ROS topics", could we call it `subscribeToAllROSTopics`?